theseus/optimizer/nonlinear/tests/test_backwards.py

```diff
@@ -0,0 +1,118 @@
+    noise = torch.randn((1, num_points)) * noise_factor
+    data_y = a * data_x.square() + b + noise
+    return data_x, data_y
+
+
+num_points = 10
+data_x, data_y = generate_data(num_points)
+
+x = th.Variable(data_x.requires_grad_(), name="x")
+y = th.Variable(data_y.requires_grad_(), name="y")
+a = th.Vector(1, name="a")
+b = th.Vector(1, name="b")
+
+
+def quad_error_fn(optim_vars, aux_vars):
+    a, b = optim_vars
+    x, y = aux_vars
+    est = a.data * x.data.square() + b.data
+    err = y.data - est
+    return err
+
+
+optim_vars = [a, b]
+aux_vars = [x, y]
+cost_function = th.AutoDiffCostFunction(
+    optim_vars,  # type: ignore
+    quad_error_fn,
+    num_points,
+    aux_vars=aux_vars,
+    name="quadratic_cost_fn",
+)
+objective = th.Objective()
+objective.add(cost_function)
+optimizer = th.GaussNewton(
+    objective,
+    max_iterations=15,
+    step_size=0.5,
+)
+
+theseus_inputs = {
+    "a": 2 * torch.ones((1, 1)).requires_grad_(),
+    "b": torch.ones((1, 1)).requires_grad_(),
+    "x": data_x,
+    "y": data_y,
+}
+theseus_optim = th.TheseusLayer(optimizer)
+
+
+def test_backwards():
+    # First we use numdifftools to numerically compute the gradient
+    # the optimal a w.r.t. the x part of the data
+    def fit_x(data_x_np):
+        theseus_inputs["x"] = (
+            torch.from_numpy(data_x_np).float().clone().requires_grad_().unsqueeze(0)
+        )
+        updated_inputs, info = theseus_optim.forward(
+            theseus_inputs, track_best_solution=True, verbose=False
+        )
+        return updated_inputs["a"].item()
+
+    data_x_np = data_x.detach().clone().numpy()
+    dfit_x = nd.Gradient(fit_x)
+    da_dx_numeric = torch.from_numpy(dfit_x(data_x_np)).float()
+
+    theseus_inputs["x"] = data_x
+    updated_inputs, info = theseus_optim.forward(
+        theseus_inputs,
+        track_best_solution=True,
+        verbose=False,
+        backward_mode=th.BackwardMode.FULL,
+    )
+    da_dx_full = torch.autograd.grad(updated_inputs["a"], data_x, retain_graph=True)[
+        0
+    ].squeeze()
+    assert torch.allclose(da_dx_numeric, da_dx_full, atol=1e-3)
+
+    # TODO: Debug a little more
+    # updated_inputs, info = theseus_optim.forward(
+    #     theseus_inputs,
+    #     track_best_solution=True, verbose=False,
+    #     backward_mode=th.BackwardMode.IMPLICIT,
+    # )
+    # da_dx_implicit = torch.autograd.grad(
+    #     updated_inputs['a'], data_x,
+    #     retain_graph=True)[0].squeeze()
+    # EPS=1e-3
+    # assert torch.allclose(da_dx_numeric, da_dx_implicit, atol=EPS)
+
+    updated_inputs, info = theseus_optim.forward(
+        theseus_inputs,
+        track_best_solution=True,
+        verbose=False,
+        backward_mode=th.BackwardMode.TRUNCATED,
+        backward_num_iterations=5,
+    )
+    da_dx_truncated = torch.autograd.grad(
+        updated_inputs["a"], data_x, retain_graph=True
+    )[0].squeeze()
+    assert torch.allclose(da_dx_numeric, da_dx_truncated, atol=1e-1)
```

Comment:
This `atol` seems high. However, since the gradient is actually a constant factor of the numeric gradient (I confirmed this), does it make sense to change this test? For example, something like

```python
diff_ratio = da_dx_numeric / da_dx_truncated
assert diff_ratio.max() - diff_ratio.min()  < TOLERANCE
```

BTW, for implicit, the ratio seems to be exactly = 2. Is this particular of this problem or have you observed this in other problems as well? Also, for truncated, the relative difference wrt reference seems to decrease exponentially in the number of iterations. You may had noticed this already, but just thought I mentioned it, since it could be useful/interesting.

Reply:
Fixed the scaling issue in https://github.com/facebookresearch/theseus/issues/39 and updated the tests to have `atol=1e-4` for both checks